File: algorithm-java/src/main/java/com/leetcode/StackSort.java
package com.leetcode;

import java.util.Stack;

/**
 * sort stack
 * 
 * @author clx 4/18/2019
 */
public class StackSort {

	/**
	 * sort
	 * 
	 * @param srcStack
	 * @return
	 */
	public static void sortStack(Stack<Integer> srcStack) {
		if (srcStack == null || srcStack.size() == 0) {
			return;
		}
		Stack<Integer> tmpStack = new Stack<>();
		tmpStack.push(srcStack.pop());
		int top;
		while (!srcStack.isEmpty()) {
			top = srcStack.pop();
			while (!tmpStack.isEmpty() && top > tmpStack.peek()) {
				srcStack.push(tmpStack.pop());
			}
			tmpStack.push(top);
		}
		while (!tmpStack.isEmpty()) {
			srcStack.push(tmpStack.pop());
		}
	}
}

File: algorithm-go/src/leetcode/stack_sort.go
package leetcode

import (
	"leetcode/common"
)

// StackSort sort stack
func StackSort(stack *common.StackArray) {
	if stack == nil {
		return
	}
	cacheStack := common.NewStack()
	v := stack.Pop()
	_ = cacheStack.Push(v)
	var top interface{}
	for !stack.IsEmpty() {
		top = stack.Pop()
		for !cacheStack.IsEmpty() && typeConvert(top) > typeConvert(cacheStack.Peek()) {
			stack.Push(cacheStack.Pop())
		}
		cacheStack.Push(top)
	}
	for !cacheStack.IsEmpty() {
		stack.Push(cacheStack.Pop())
	}
}

func typeConvert(value interface{}) int {
	if value != nil {
		switch value.(type) {
		case int:
			if v, ok := value.(int); ok {
				return v
			}
		default:
			return -1
		}
	}
	return -1
}
